### DISCUSSION TITLE
Can I convert a delegate to a C# function pointer? (eg. delegate* managed<void>)

### DISCUSSION THREAD

**Discussion by cyraid** [dotnet/runtime#84026] [Q&A] [answered]
As the title suggests, is there a way to convert a delegate into a **managed** function pointer? (https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers) .. I'm using them, and they work great, but I don't want to use "&func" in an unsafe context, I was hoping the underlying method call could access it, and I can just provide a delegate.

**Comment by DaZombieKiller** [comment, marked as answer]
You can currently only do this if the delegate has no target, is not a `DynamicMethod` and is static (as far as I know):
```cs
static bool TryGetFunctionPointer(Delegate d, out void* pointer)
{
    ArgumentNullException.ThrowIfNull(d);
    var method = d.Method;

    if (d.Target is {} || !method.IsStatic || method is DynamicMethod)
    {
        pointer = null;
        return false;
    }

    pointer = (void*)method.MethodHandle.GetFunctionPointer();
    return true;
}
```
[SharpLab Sample](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyA6AJRgDMAbGMDASwh5mAUQC2/DAG5q1fAGcAhhxjUA3tXpb6AExg8A5oowwAVAB4iKAHz0OABxlVt9AG6Ko9RX0F56AXno8GAB3RjISAAoRBEUxex4AWRgMAAsIHQBKJxdNbX4OekiAQgAVKABPAHEUogBWASEABQh+PBMoSO9GvDR6CBwMelNIq1NTTIAyB0zMvK0NZxdtJgBOSIAiADFFfh4degwIegMUu3wfIXp7VvaYKGYN7PnlogB2HO0AXxeXh0jnlQflRqEwkIwUPQYnEEjBkmkMgD1C81psABL6LgQPohaBcHTFJ5OYGgsjghTKGD0YAQCBcejlaq1Bq+FptDqRAAi+hgRhMuj6AyGY2utw6cyoixcAEEoAYcGIYO0AHI4LhcGJgGD2HrMUqpKAQEIASQ4qvVkSyny07k8ioRB0COmY8PSOicLwKRWdpQ8pyG/Hk9DUX3oBAI9GK9rdzGN8gAyhhjPwwGGI9GMvRA/RORU8HEU66MhKXFLllobuz7gEgmquNbXm87IouPIYA2SUsK2Lq4FRigJhnnUWdGjFHgdDxmDUMPUemy7p1AS53odcO3qMCgA==)

**Reply by cyraid** [reply]
It totally works! Thank you very much! Now I don't have to do a &Method and unsafe! :) .. It's too bad it doesn't work with lambda but that's quite alright.

Thanks again man!
